core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java

```diff
@@ -238,6 +253,18 @@ public HttpRemoteTask(
             }
             maxUnacknowledgedSplits = getMaxUnacknowledgedSplitsPerTask(session);
 
+            this.guaranteedSplitsPerRequest = getRemoteTaskGuaranteedSplitsPerRequest(session);
+            this.maxRequestSize = getMaxRemoteTaskRequestSize(session).toBytes();
+            this.requestSizeHeadroom = getRemoteTaskRequestSizeHeadroom(session).toBytes();
+            this.splitBatchSize = maxUnacknowledgedSplits;
+            long numOfPartitionedSources = planFragment.getPartitionedSources().size();
+            // Currently it supports only when there is one partitioned source.
+            this.enableAdaptiveTaskRequestSize = numOfPartitionedSources == 1 ? isEnableAdaptiveTaskRequestSize(session) : false;
```

Comment:
It would be great if it worked for any number / type of sources. Otherwise some queries (such as collocated joins) might still experience the same problem. Would it be possible to have something like `targetSplitsPerSource` and adjust it according to request size?

Reply:
It feels like you would need to change quite a bit, but maybe it is more incremental than it seems. Do you think it is worth having two increments?